File: bubble_sort/bubble.c
/*
  Bubble Sorting Algorithm on C
*/

#include <stdio.h>
#include <stdlib.h>
#include <malloc.h>

void bubble_sort(int *array, int len);
void swap(int *a, int *b);
void print_array(char *message, int *array, int len);

int main(int argc, char *argv) {
    printf("Bubble Sorting Algorithm\n\n");
    FILE *f = fopen("array.txt", "r+");
    int len = 0, *array = NULL;
    for (len = 0; !feof(f); len++) {
        array = (int *)realloc(array, sizeof(int)*(len + 1));
        fscanf(f, "%d\n", &array[len]);
    }
    print_array("Before:", array, len);
    bubble_sort(array, len);
    print_array("After: ", array, len);
    free(array);
    return 0;
}

void bubble_sort(int *array, int len) {
    for (int i = 0; i < len; i++)
      for (int j = 0; j < len-i-1; j++)
        if (array[j] > array[j+1])
          swap(&array[j], &array[j+1]);
}

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

void print_array(char *message, int *array, int len) {
    printf("%s", message);
    for (int i = 0; i < len; i++)
        printf(" %d", array[i]);
    printf("\n");
}

File: bubble_sort/bubble.rb
=begin
  Bubble Sorting Algorithm on Ruby
=end

class Bubble
  def bubble(a)
    for i in 0..a.length-1
      for j in 0..(a.length-i-2)
        a[j], a[j+1] = a[j+1], a[j] if (a[j] > a[j+1])
      end
    end
  end
end

class Start
  def initialize
    print "Bubble Sorting Algorithm\n\n"
    a = []
    File.open("array.txt").each_with_index { |line, i|
      a[i] = line.to_i
    }
    print_array("Before:", a)
    Bubble.new.bubble(a)
    print_array("After: ", a)
  end

  def print_array(message, a)
      print message
      a.each { |x| print " #{x}"}
      print "\n"
  end
end

Start.new
